src/stream.rs

```diff
@@ -0,0 +1,216 @@
+use std::{pin::Pin, task::Poll};
+
+use futures_core::{Future, Stream, TryStream};
+use pin_project_lite::pin_project;
+
+use crate::{
+    backoff::Backoff,
+    future::{rt_sleeper, Sleeper},
+};
+
+/// Applies a [`Backoff`] policy to a [`Stream`]
+///
+/// After any [`Err`] is emitted, the stream is paused for [`Backoff::next_backoff`]. The
+/// [`Backoff`] is [`reset`](`Backoff::reset`) on any [`Ok`] value.
+///
+/// If [`Backoff::next_backoff`] returns [`None`] then the backing stream is given up on, and closed.
+#[cfg(any(feature = "tokio", feature = "async-std"))]
+pub fn backoff<S: TryStream, B: Backoff>(
```

Comment:
Ideally this'd be an extension method on `TryStream`, but that isn't allowed for `impl Trait` return values.